cpp/lib/include/Subfields.h

```diff
@@ -21,46 +21,71 @@
 
 
+#include <algorithm>
+#include <iostream>
+#include <PriorityQueue.h>
 #include <unordered_map>
 #include <string>
 #include <vector>
 
 
+struct Subfield {
+    char code_;
+    std::string value_;
+public:
+    Subfield(const char code, const std::string &value) : code_(code), value_(value) {}
+};
+
+struct SubfieldCodeLessThan {
+    bool operator() (const Subfield &lhs, const Subfield &rhs) { return lhs.code_ < rhs.code_; }
+    bool operator() (const Subfield &lhs, const char &rhs) { return lhs.code_ < rhs; }
+    bool operator() (const char &lhs, const Subfield &rhs) { return lhs < rhs.code_; }
+};
+
+struct CompareSubfieldCode {
+    char code_;
+public:
+    explicit CompareSubfieldCode(const char code) : code_(code) {}
+    bool operator() (const Subfield &subfield) { return subfield.code_ == code_; }
+};
+
 /** \class Subfields
  *  \brief Encapsulates the subfields of a MARC-21 data field.
  */
 class Subfields {
     char indicator1_, indicator2_;
-    std::unordered_multimap<char, std::string> subfield_code_to_data_map_;
+
+    std::vector<Subfield> subfields_;
 public:
-    typedef std::unordered_multimap<char, std::string>::const_iterator ConstIterator;
-    typedef std::unordered_multimap<char, std::string>::iterator Iterator;
+    typedef std::vector<Subfield>::const_iterator ConstIterator;
+    typedef std::vector<Subfield>::iterator Iterator;
 public:
-    Subfields() {}
-    Subfields(const char indicator1, const char indicator2): indicator1_(indicator1), indicator2_(indicator2) {}
+    Subfields() : indicator1_('\0'), indicator2_('\0') { }
+    Subfields(const char indicator1, const char indicator2)
+        : indicator1_(indicator1), indicator2_(indicator2) { }
 
     /** \brief Parses a binary MARC-21 field. */
     explicit Subfields(const std::string &field_data);
 
-    bool empty() const { return subfield_code_to_data_map_.empty(); }
-    size_t size() const { return subfield_code_to_data_map_.size(); }
+    bool empty() const { return subfields_.empty(); }
+    size_t size() const { return subfields_.size(); }
     char getIndicator1() const { return indicator1_; }
     void setIndicator1(const char indicator1) { indicator1_ = indicator1; }
     char getIndicator2() const { return indicator2_; }
     void setIndicator2(const char indicator2) { indicator2_ = indicator2; }
     bool hasSubfield(const char subfield_code) const
-        { return subfield_code_to_data_map_.find(subfield_code) != subfield_code_to_data_map_.end(); }
+        { return std::find_if(begin(), end(), CompareSubfieldCode(subfield_code)) != end(); }
 
     /** \return True, if a subfield with subfield code "subfield_code" and contents "value" exists, else false. */
     bool hasSubfieldWithValue(const char subfield_code, const std::string &value) const;
 
     /** \return The bounds of the range of entries that have a subfield code of "subfield_code". */
     std::pair<ConstIterator, ConstIterator> getIterators(const char subfield_code) const
```

Comment:
I think "std::pair<ConstIterator, ConstIterator>" should be replaced with "std::pair<const_iterator, const_iterator>".